server/grammar/YmlLexer.g4

```diff
@@ -0,0 +1,153 @@
+CLASSPROPERTIES: 'classProperties';
+EXTERN: 'extern';
+COMPLETE: 'complete';
+SYNONYM: 'synonym';
+OVERRIDE: 'override';
+ARGS: 'args';
+IF_EXPR: 'ifExpr';
+IF: 'if';
+THEN: 'then';
+ELSE: 'else';
+ENUM: 'enum';
+DO: 'do';
+TRY: 'try';
+CATCH: 'catch';
+FOREACH: 'foreach';
+FORALL: 'forall';
+IN: 'in';
+AS: 'as';
+RETURN: 'return';
+LOCAL: 'local';
+TRUE: 'true';
+FALSE: 'false';
+SWITCH: 'switch';
+CASE: 'case';
+DEFAULT: 'default';
+BREAK: 'break';
+STATIC: 'static';
+WHILE: 'while';
+FOR: 'for';
+
+//Symbols
+SEMICOLON: ';';
+COMMA: ',';
+COLON: ':';
+DOT: '.';
+MULTIVALUED_EXPRESSION: '>>';
+FIELD_INTRO: '-->';
+REPLACE_FIELD_VALUE_INTRO: '==>';
+REMOVE_FIELD: '---';
+ADD_FIELD: '+++';
+INLINE_DECL_INTRO: '->';
+OPEN_PAR: '(';
+CLOSE_PAR: ')';
+OPEN_BRACKET: '[';
+CLOSE_BRACKET: ']';
+OPEN_BRACE: '{';
+CLOSE_BRACE: '}';
+SINGLE_QUOTE: '\'';
+BACKSLASH: '\\';
+QUESTION_MARK: '?';
+AT: '@';
+
+//Comparison operator
+EQUAL_COMP: '==';
+NOT_EQUALS: '!=';
+LESS_OR_EQUAL: '<=';
+GREATE_OR_EQUAL: '>=';
+STRICT_LESS: '<';
+STRICT_GREAT: '>';
+
+//Condition combination
+COND_AND: '&&';
+COND_OR: '||';
+
+//Assignment operators
+EQUAL_ASSIGNMENT: '=';
+MULTIVALUED_ASSIGNMENT: ':=';
+
+OPEN_GRANULE: BACKSLASH OPEN_PAR;
+CLOSE_GRANULE: BACKSLASH CLOSE_PAR;
+
+fragment MINUS: '-';
+fragment PLUS: '+';
+fragment DIVIDE: '/';
+fragment TIMES: '*';
+
+OPERATOR: MINUS | PLUS | DIVIDE | TIMES;
+
+fragment NUMBER: [0-9];
+fragment LETTER: [\p{Letter}_];
+
+fragment ALPHANUM: LETTER | [0-9];
+fragment D_LETTER: 'd';
+
+DATE:
+    SINGLE_QUOTE D_LETTER COLON NUMBER NUMBER NUMBER NUMBER MINUS NUMBER NUMBER MINUS NUMBER NUMBER SINGLE_QUOTE
+;
+
+fragment ESCAPED_QUOTE: '\\"';
+STRING: '"' ( ESCAPED_QUOTE | ~('\n' | '\r'))*? '"';
+fragment TRIPLE_QUOTE: '"""';
+DOCUMENTATION: TRIPLE_QUOTE (.*?) TRIPLE_QUOTE;
+
+//We store the whitespace in hidden channel to take back the ones around comments
+WS: [ \r\t\n]+ -> channel(HIDDEN);
+
+DOUBLE: MINUS? NUMBER+ (DOT NUMBER+)?;
+INTEGER: NUMBER+;
+
+/*
+ * Colon is OK inside a YMLID
+ * Simple colon has no specific meaning, however:
+ * − double colons are used to define a class' method;
+ * − triple colons are used as a pointer towards a class' method or attribute.
+ */
+YMLID: ID (COLON ID)* (COLON? COLON? COLON ID)?;
+ID: ALPHANUM* LETTER ALPHANUM*;
+
+fragment MULTILINE_COMMENT_START: '/*';
+fragment MULTILINE_COMMENT_END: '*/';
+
+PREPROCESSING: '@' ~[\n\r]* -> channel(HIDDEN);
+LINE_COMMENT: '//' ~[\n\r]* -> channel(HIDDEN);
+MULTILINE_COMMENT:
+    MULTILINE_COMMENT_START .*? MULTILINE_COMMENT_END -> channel(HIDDEN)
+;
+
+mode DECLARATION_FILE;
+
+DECL_FILE_PREPROCESSING: PREPROCESSING -> channel(HIDDEN);
+DECL_FILE_LINE_COMMENT: LINE_COMMENT -> channel(HIDDEN);
+DECL_FILE_MULTILINE_COMMENT: MULTILINE_COMMENT -> channel(HIDDEN);
+
+NAME: '.'? ALPHANUM+ ('-' ALPHANUM+)*;
+
+// Could be more complicated, but should be enough in a first place.
+FILE_DECLARATION: ('./'? | '../'*) NAME ('/' NAME)* ('.' NAME)*;
```

Comment:
I don’t quite get the purpose of the final `('.' NAME)*` :thinking: Especially since there seems to be another (optional) `.` in `NAME`. Can you show me an example of line that needs this to be recognized?

Reply:
Lol, could be `DECL_FILE_FILE_DECLARATION` but :nauseated_face:

Reply:
What if we just looked for something equivalent to the regex `[^@()]+`? Provided that comments take precedence over this mess.

Reply:
Consecutive dots are OK, you know.

```
    $ find Music/ -type f -wholename '*....*'
[…]
Music/mesarthim/_.-_-..._..._._-._-.-/flac/05_-_......flac
[…]
Music/drudkh/they_often_see_dreams_about_the_spring/flac/02_-_u_dakhiv_irzhavim_kolossyu....flac
[…]
Music/wedrujacy_wiatr/tam_gdzie_miesiac_oplakuje_swit/flac/02_-_tam_gdzie_miesiac_oplakuje_swit....flac
[…]
```

### Trivia

The `_` at the beginning of the `.-_-..._..._._-._-.-` album was manually added to prevent it from becoming a hidden file.

Reply:
`mesarthim` seems great.

Reply:
Sorry, I think I don't understand what should be the next step on this issue.

Reply:
Well, I think your pattern is – ironically – needlessly strict and complicated.

I might make errors because Antlr noob, but intuitively I’d write:

```
NAME: ('-' | '.' | ALPHANUM)+;
FILE_DECLARATION: NAME ('/' NAME)*;
```

Reply:
Oh and `_` should be accepted too, right?

Reply:
I had to improve a little your rules, but it was great.
`_` is already in ALPHANUM.

Reply:
> `_` is already in ALPHANUM.

:open_mouth: × 10